### DISCUSSION TITLE
Issue with Multiple Columns in Custom PostgreSQL Function

### DISCUSSION THREAD

**Discussion by Matignaciom** [supabase/supabase#28325] [Questions] [answered]
I'm encountering an issue with a PostgreSQL function that retrieves data based on a user ID and a list of columns. When querying multiple columns, the function only returns the phone number, rather than displaying all requested columns separated by comma and space. The function works correctly for a single column but fails to concatenate multiple columns as expected.

**Here is the code for the function**:

```
DROP FUNCTION IF EXISTS obtener_usuarios(TEXT, TEXT);

CREATE OR REPLACE FUNCTION obtener_usuarios(user_id TEXT, columns TEXT)
RETURNS TABLE(result TEXT) AS $$
DECLARE
    query TEXT;
    valid_columns TEXT[] := ARRAY[
        'id', 'email', 'nombre', 'direccion', 'telefono', 'fecha_nacimiento', 
        'tipo_perfil', 'iglesia_id', 'contrasena', 'fecha_creacion', 'habilitada'
    ];
    column_list TEXT[];
    col TEXT;
    column_count INT;
BEGIN
    column_list := string_to_array(columns, ',');
    column_count := array_length(column_list, 1);
    
    FOREACH col IN ARRAY column_list
    LOOP
        col := trim(both ' ' from col);
        
        IF NOT col = ANY(valid_columns) THEN
            RAISE EXCEPTION 'Invalid column name: %', col;
        END IF;
    END LOOP;
    
    IF column_count = 1 THEN
        query := 'SELECT ' || quote_ident(trim(both ' ' from column_list[1])) || '::TEXT FROM usuarios WHERE id = ' || quote_literal(user_id);
    ELSE
        query := 'SELECT ' ||
                 '  string_agg(value, '', '') ' ||
                 'FROM ( ' ||
                 '  SELECT ' ||
                 '    concat_ws('' '', ' ||
                 '      ' || string_agg('COALESCE(' || quote_ident(trim(both ' ' from col)) || ', '''')', ', ') || 
                 '    ) AS value ' ||
                 '  FROM usuarios ' ||
                 '  WHERE id = ' || quote_literal(user_id) ||
                 ') sub';
    END IF;

    RETURN QUERY EXECUTE query;
END;
$$ LANGUAGE plpgsql;

-- Example calls:
-- Single column
SELECT * FROM obtener_usuarios('bbdaa840-f2b6-42af-b945-90d780b835b3', 'nombre');

-- Multiple columns
SELECT * FROM obtener_usuarios('bbdaa840-f2b6-42af-b945-90d780b835b3', 'nombre, direccion, telefono');
```

**Example calls:**

`Single column: SELECT * FROM obtener_usuarios('bbdaa840-f2b6-42af-b945-90d780b835b3', 'nombre');`

`Multiple columns: SELECT * FROM obtener_usuarios('bbdaa840-f2b6-42af-b945-90d780b835b3', 'nombre, direccion, telefono');`

**The issue is that** only the phone number is shown for multiple columns, and the concatenation of values isn't working as expected. Any help to resolve this issue would be greatly appreciated!

**Comment by GaryAustin1** [comment]
This appears to just use the last column name from your loop which is stored in col and would be telefono.
![image](https://github.com/user-attachments/assets/9a2ec746-aebc-41e1-bbd1-27fbf9f32f82)


I don't have suggestions on how to do what you want, but you might ask an AI.

**Reply by Matignaciom** [reply]
I'm learning certain things in Supabase, and what I want to do is create a function that allows me to query a single column or multiple columns from a table (the users table in this case). When I query a single column, it works perfectly, but when I try with multiple columns, it doesn't work. :( 

I tried asking GPT, but I still can’t solve it...

**Reply by GaryAustin1** [reply]
This is not a Supabase question.  This is a Postgres SQL question.  It is fine to ask here, but there are many places to search for Postgres help.

I assume you have some reason for doing this in a postgres function versus just using a client and change .select(*) to .select('commaSeperatedColumnString').

**Reply by GaryAustin1** [reply, marked as answer]
Don't know if this is correct or meets your needs, but I use Cody AI in my IDE (uses Claude 3.5 Sonnet for the engine) and it generated this when I asked it to fix your function to return the columns in the input list...

```SQL
CREATE OR REPLACE FUNCTION obtener_usuarios(user_id TEXT, columns TEXT)
RETURNS TABLE(result JSON) AS $$
DECLARE
    query TEXT;
    valid_columns TEXT[] := ARRAY[
        'id', 'email', 'nombre', 'direccion', 'telefono', 'fecha_nacimiento', 
        'tipo_perfil', 'iglesia_id', 'contrasena', 'fecha_creacion', 'habilitada'
    ];
    column_list TEXT[];
    col TEXT;
BEGIN
    column_list := string_to_array(columns, ',');
    
    FOREACH col IN ARRAY column_list
    LOOP
        col := trim(both ' ' from col);
        
        IF NOT col = ANY(valid_columns) THEN
            RAISE EXCEPTION 'Invalid column name: %', col;
        END IF;
    END LOOP;
    
    query := 'SELECT row_to_json(row) FROM (SELECT ' || 
             array_to_string(array(SELECT quote_ident(trim(both ' ' from col)) FROM unnest(column_list) AS col), ', ') || 
             ' FROM usuarios WHERE id = ' || quote_literal(user_id) || ') row';
    
    RETURN QUERY EXECUTE query;
END;
$$ LANGUAGE plpgsql;
```
https://sourcegraph.com/cody

**Reply by Matignaciom** [reply]
Thank you very much for the help
